File: engine/prioritization.js
// Prioritization heuristics for deciding what must stay in the window.

import { SourceKind, RetrievedItem, estimateTokens } from "./models.js";

const DEFAULT_KIND_BASE = {
  [SourceKind.PREFERENCE]: 5.0,
  [SourceKind.TRAJECTORY]: 3.0,
  [SourceKind.HISTORY]: 2.5,
  [SourceKind.CODE]: 2.0,
  [SourceKind.FILE]: 1.5,
};

export class PriorityConfig {
  constructor({
    kindBase = DEFAULT_KIND_BASE,
    recencyWeight = 0.2,
    priorityWeight = 1.0,
    lengthPenaltyWeight = 0.1,
    maxLengthTokens = 800,
  } = {}) {
    this.kindBase = { ...DEFAULT_KIND_BASE, ...kindBase };
    this.recencyWeight = recencyWeight;
    this.priorityWeight = priorityWeight;
    this.lengthPenaltyWeight = lengthPenaltyWeight;
    this.maxLengthTokens = maxLengthTokens;
  }
}

export function prioritize(items, config = null) {
  const cfg = config || new PriorityConfig();
  const ranked = [];
  for (const item of items) {
    const chunk = item.chunk;
    const kindBase = cfg.kindBase[chunk.kind] ?? 1.0;
    const recency = recencyScore(chunk);
    const lengthPenalty = Math.min(1, estimateTokens(chunk.content) / cfg.maxLengthTokens);
    const priority = chunk.priority / 10.0;
    const score =
      item.score +
      kindBase * 0.1 +
      recency * cfg.recencyWeight +
      priority * cfg.priorityWeight -
      lengthPenalty * cfg.lengthPenaltyWeight;
    ranked.push(new RetrievedItem({ chunk, score, reason: item.reason }));
  }
  ranked.sort((a, b) => b.score - a.score);
  return ranked;
}

function recencyScore(chunk, halfLifeDays = 7.0) {
  const now = Date.now() / 1000;
  const ageDays = Math.max(0, (now - (chunk.timestamp || now)) / 86400);
  return Math.exp(-ageDays / halfLifeDays);
}

File: python/local_context_engine/prioritization.py
"""Prioritization heuristics for deciding what must stay in the window."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, Iterable, List, Optional

from .models import Chunk, RetrievedItem, SourceKind


@dataclass
class PriorityConfig:
    """Weights applied after retrieval to re-order candidates."""

    kind_base: Dict[SourceKind, float] = field(default_factory=lambda: {
        SourceKind.PREFERENCE: 5.0,
        SourceKind.TRAJECTORY: 3.0,
        SourceKind.HISTORY: 2.5,
        SourceKind.CODE: 2.0,
        SourceKind.FILE: 1.5,
    })
    recency_weight: float = 0.2
    priority_weight: float = 1.0
    length_penalty_weight: float = 0.1
    max_length_tokens: int = 800


def prioritize(
    items: Iterable[RetrievedItem],
    config: Optional[PriorityConfig] = None,
) -> List[RetrievedItem]:
    """Combine retrieval score with source/recency/importance heuristics."""
    config = config or PriorityConfig()
    ranked: List[RetrievedItem] = []
    for item in items:
        chunk = item.chunk
        kind_base = config.kind_base.get(chunk.kind, 1.0)
        recency = _recency_score(chunk)
        length_penalty = min(1.0, _approx_tokens(chunk.content) / config.max_length_tokens)
        priority = chunk.priority / 10.0
        score = (
            item.score
            + kind_base * 0.1
            + recency * config.recency_weight
            + priority * config.priority_weight
            - length_penalty * config.length_penalty_weight
        )
        ranked.append(RetrievedItem(chunk=chunk, score=score, reason=item.reason))
    ranked.sort(key=lambda item: item.score, reverse=True)
    return ranked


def _recency_score(chunk: Chunk, half_life_days: float = 7.0) -> float:
    import math
    import time

    now = time.time()
    age_days = max(0.0, (now - (chunk.timestamp or now)) / 86400.0)
    return math.exp(-age_days / half_life_days)


def _approx_tokens(text: str) -> int:
    from .models import estimate_tokens

    return estimate_tokens(text)
